JDBC Step4 띠용

jye326 opened the pull request:
벌써 마지막이네요 ㅎㅎ 이번에도 자유롭게 리뷰 부탁드려요~

jye326 committed:
feat: UserService 인터페이스 분리 및 트랜잭션 관리 개선

- UserService를 인터페이스로 분리하여 책임 분리
- AppUserService와 TxUserService 구현체 추가
- TxUserService에 TransactionManager 적용으로 트랜잭션 처리 캡슐화
- 코드 가독성과 유지보수성 향상

jye326 committed:
refactor: 트랜잭션 관리 개선 및 중복 코드 제거

- TransactionSynchronizationManager에 자원 바인딩 및 해제 로직 추가
- DataSourceUtils의 releaseConnection 개선으로 안전한 리소스 해제
- JdbcTemplate 및 TransactionManager에서 중복 코드 제거 및 자원 관리 개선
- UserDao 및 UserHistoryDao의 Connection 기반 메서드 제거로 사용되지 않는 메서드 정리
- AppUserService 및 TxUserService로 트랜잭션 분리 및 예외 메시지 명확화

jye326 committed:
refactor: releaseConnection 메서드의 트랜잭션 바인딩 검증 로직 개선

- TransactionSynchronizationManager의 isBound 메서드 활용으로 코드 간결화
- 불필요한 null 체크 로직 제거 및 리소스 해제 처리 안정

20HyeonsuLee commented on jdbc/src/main/java/com/interface21/transaction/support/TransactionSynchronizationManager.java:
```diff
@@ -1,23 +1,41 @@
+        if (!isBound(key)) {
+            throw new IllegalStateException("기존에 바인딩 된 연결이 존재하지 않습니다.");
+        }
+        Map<DataSource, Connection> map = resources.get();
+        Connection removed = map.remove(key);
+        return removed;
```
인텔리제이에서 경고를 발생시키고 있네요.

<img width="361" height="80" alt="image" src="https://github.com/user-attachments/assets/f06d13ab-d552-4169-8c39-a245b51bf559" />

요렇게도 가능할 것 같습니다!
```suggestion
        return map.remove(key);
```

jye326 replied:
remove 직후에 ThreadLocal 을 제거하는 로직이 추가되어 인라인 처리를 안 해도 될 것 같아요!

20HyeonsuLee commented on jdbc/src/main/java/com/interface21/transaction/TransactionManager.java:
```diff
@@ -14,19 +16,28 @@ public TransactionManager(DataSource dataSource) {
-            throw new RuntimeException("커넥션 오류 발생", e);
+            throw new RuntimeException("SQL 오류 발생", e);
+        } finally {
+            try {
+                connection.setAutoCommit(true);
+            } catch (SQLException e) {
+                throw new RuntimeException(e);
+            }
+            DataSourceUtils.releaseConnection(connection, dataSource);
+            TransactionSynchronizationManager.unbindResource(dataSource);
         }
```
요 리뷰 한번 확인해주시죠!

jye326 replied:
아이고 큰일날 뻔 했네요;;

20HyeonsuLee commented on jdbc/src/main/java/com/interface21/transaction/TransactionManager.java:
```diff
@@ -14,19 +16,28 @@ public TransactionManager(DataSource dataSource) {
-            throw new RuntimeException("커넥션 오류 발생", e);
+            throw new RuntimeException("SQL 오류 발생", e);
+        } finally {
+            try {
+                connection.setAutoCommit(true);
+            } catch (SQLException e) {
+                throw new RuntimeException(e);
+            }
```
`e.addSuppressed(ex);`를 이용하면 예외를 덮어씌우는 문제를 더 간단하게 해결할 수 있을 것 같습니다!

```java
try {
            connection.setAutoCommit(false);
            userService.changePassword(id, newPassword, createBy);
            connection.commit();
        } catch (Exception e) {
            try {
                connection.rollback();
            } catch (SQLException ex) {
                e.addSuppressed(ex);
            }
            throw new DataAccessException(e);
        } finally {
            DataSourceUtils.releaseConnection(connection, dataSource);
            TransactionSynchronizationManager.unbindResource(dataSource);
        }
```

jye326 replied:
그렇네요! 반영했습니다~

20HyeonsuLee reviewed (changes requested):
안녕하세요 띠용~ 벌써 마지막 step이네요.

정말 빠르게 구현해주셨군요👍👍 테코톡이 오늘이셨는데.. 대단해요

코드레빗의 리뷰중 중요한 부분이 있는 것 같아서 요거만 반영해주시죠!

jye326 commented on jdbc/src/main/java/com/interface21/transaction/support/TransactionSynchronizationManager.java:
```diff
@@ -1,23 +1,41 @@
 import java.util.Map;
+import javax.sql.DataSource;
 
 public abstract class TransactionSynchronizationManager {
 
-    private static final ThreadLocal<Map<DataSource, Connection>> resources = new ThreadLocal<>();
+    private static final ThreadLocal<Map<DataSource, Connection>> resources = ThreadLocal.withInitial(HashMap::new);
 
-    private TransactionSynchronizationManager() {}
+    private TransactionSynchronizationManager() {
+    }
 
     public static Connection getResource(DataSource key) {
-        return null;
+        Map<DataSource, Connection> map = resources.get();
+        return map.get(key);
     }
 
     public static void bindResource(DataSource key, Connection value) {
+        Map<DataSource, Connection> map = resources.get();
+        if (map.containsKey(key)) {
+            throw new IllegalStateException("이미 커넥션이 존재합니다.");
+        }
+        map.put(key, value);
     }
 
     public static Connection unbindResource(DataSource key) {
-        return null;
+        if (!isBound(key)) {
+            throw new IllegalStateException("기존에 바인딩 된 연결이 존재하지 않습니다.");
+        }
+        Map<DataSource, Connection> map = resources.get();
+        Connection removed = map.remove(key);
+        return removed;
+    }
+
+    public static boolean isBound(DataSource key){
+        Map<DataSource, Connection> map = resources.get();
+        return map.containsKey(key);
     }
```
ThreadLocal 자체도 제거를 해야 하는구나!

jye326 committed:
refactor: releaseConnection 호출 순서 수정

- 트랜잭션 리소스 해제 순서 명확화 및 안정성 강화

jye326 committed:
refactor: 트랜잭션 자원 해제 로직 개선 및 불필요한 import 제거

- TransactionSynchronizationManager에 releaseThreadLocalMap 메서드 추가로 자원 해제 처리 간결화
- 트랜잭션 리소스 해제 시 ThreadLocal에서 빈 맵 제거로 메모리 낭비 방지
- AppUserService 및 UserService에서 사용되지 않는 Connection import 제거로 코드 정리 및 가독성 향상

jye326 committed:
refactor: 트랜잭션 롤백 및 autoCommit 복원 로직 개선

- 트랜잭션 롤백 시 발생한 SQLException을 suppressed로 추가하여 디버깅 정보 강화
- 트랜잭션 종료 후 원래 autoCommit 상태로 복원하도록 로직 수정
- DataAccessException 메시지 구체화 및 명확성 향상
